File: meeting-notetaker/python/autostart.py
#!/usr/bin/env python3
"""Start auto-join automatically when you log in — the OS-specific bit.

Each platform gets its native, no-admin mechanism, and we tell you exactly what
file we created so nothing happens behind your back:

    Windows   a hidden launcher in your Startup folder      (Start Menu\\Programs\\Startup)
    macOS     a LaunchAgent                                  (~/Library/LaunchAgents)
    Linux     a systemd --user service                       (~/.config/systemd/user)

enable() also starts it now; disable() also stops it now. Called by
`python autojoin.py enable` / `disable`.
"""

import os
import subprocess
import sys

_HERE = os.path.dirname(os.path.abspath(__file__))
AUTOJOIN = os.path.join(_HERE, "autojoin.py")

APP_NAME = "MeetingNotetaker-AutoJoin"          # Windows launcher filename
LABEL = "dev.agentcall.notetaker-autojoin"      # macOS/Linux service id


# ── Windows: a hidden .vbs in the Startup folder ────────────────────────────
def _win_startup_dir():
    return os.path.join(os.environ.get("APPDATA", ""),
                        "Microsoft", "Windows", "Start Menu", "Programs", "Startup")


def _win_vbs():
    return os.path.join(_win_startup_dir(), APP_NAME + ".vbs")


def _win_enable():
    vbs = _win_vbs()
    py = sys.executable
    # VBS string literals escape a double-quote by doubling it. This launches
    #   "<python>" "<autojoin.py>" start
    # hidden (window style 0); `start` then spawns the detached daemon and exits.
    run_arg = f'"""{py}"" ""{AUTOJOIN}"" start"'
    content = ('Set sh = CreateObject("WScript.Shell")\r\n'
               f'sh.CurrentDirectory = "{_HERE}"\r\n'
               f'sh.Run {run_arg}, 0, False\r\n')
    os.makedirs(_win_startup_dir(), exist_ok=True)
    with open(vbs, "w", encoding="utf-8") as f:
        f.write(content)
    print("  " + "✓" + " added a startup launcher so auto-join starts when you log in:")
    print(f"    {vbs}")
    return 0


def _win_disable():
    vbs = _win_vbs()
    try:
        os.remove(vbs)
        print("  removed the startup launcher — auto-join won't start on login anymore.")
    except FileNotFoundError:
        print("  (it wasn't set to start on login.)")
    except OSError as e:
        print(f"  couldn't remove {vbs}: {e}")
        return 1
    return 0


def _win_is_enabled():
    return os.path.exists(_win_vbs())


# ── macOS: a LaunchAgent plist ──────────────────────────────────────────────
def _mac_plist():
    return os.path.join(os.path.expanduser("~"), "Library", "LaunchAgents", LABEL + ".plist")


def _mac_enable():
    import plistlib
    plist = _mac_plist()
    os.makedirs(os.path.dirname(plist), exist_ok=True)
    data = {
        "Label": LABEL,
        "ProgramArguments": [sys.executable, AUTOJOIN, "run"],
        "WorkingDirectory": _HERE,
        "RunAtLoad": True,
        # Restart if it crashes, but a clean `stop` (exit 0) stays stopped.
        "KeepAlive": {"SuccessfulExit": False},
        "StandardOutPath": os.path.join(os.path.dirname(_HERE), ".notetaker", "autojoin.boot.log"),
        "StandardErrorPath": os.path.join(os.path.dirname(_HERE), ".notetaker", "autojoin.boot.log"),
    }
    with open(plist, "wb") as f:
        plistlib.dump(data, f)
    subprocess.run(["launchctl", "unload", plist], capture_output=True)   # in case it's already loaded
    r = subprocess.run(["launchctl", "load", "-w", plist], capture_output=True, text=True)
    print("  " + "✓" + f" installed a LaunchAgent: {plist}")
    if r.returncode != 0:
        print(f"  (launchctl load said: {r.stderr.strip() or r.stdout.strip()})")
        print(f"  You can load it yourself with:  launchctl load -w \"{plist}\"")
    return 0


def _mac_disable():
    plist = _mac_plist()
    subprocess.run(["launchctl", "unload", "-w", plist], capture_output=True)
    try:
        os.remove(plist)
    except FileNotFoundError:
        print("  (it wasn't set to start on login.)")
        return 0
    except OSError as e:
        print(f"  couldn't remove {plist}: {e}")
        return 1
    print("  removed the LaunchAgent — auto-join won't start on login anymore.")
    return 0


def _mac_is_enabled():
    return os.path.exists(_mac_plist())


# ── Linux: a systemd --user service ─────────────────────────────────────────
def _linux_unit():
    return os.path.join(os.path.expanduser("~"), ".config", "systemd", "user", LABEL + ".service")


def _has_systemctl():
    from shutil import which
    return which("systemctl") is not None


def _linux_enable():
    if not _has_systemctl():
        print("  systemd isn't available here. Add this to your login startup by hand:")
        print(f"    {sys.executable} {AUTOJOIN} start")
        return 1
    unit = _linux_unit()
    os.makedirs(os.path.dirname(unit), exist_ok=True)
    content = (
        "[Unit]\n"
        "Description=Meeting Notetaker auto-join (watches your calendar)\n\n"
        "[Service]\n"
        f"WorkingDirectory={_HERE}\n"
        f"ExecStart={sys.executable} {AUTOJOIN} run\n"
        "Restart=on-failure\n\n"
        "[Install]\n"
        "WantedBy=default.target\n"
    )
    with open(unit, "w", encoding="utf-8") as f:
        f.write(content)
    subprocess.run(["systemctl", "--user", "daemon-reload"], capture_output=True)
    r = subprocess.run(["systemctl", "--user", "enable", "--now", LABEL + ".service"],
                       capture_output=True, text=True)
    print("  " + "✓" + f" installed a systemd --user service: {unit}")
    if r.returncode != 0:
        print(f"  (systemctl said: {r.stderr.strip() or r.stdout.strip()})")
    print("  Tip: to run it even before you log in, enable lingering once:")
    print("    loginctl enable-linger $USER")
    return 0


def _linux_disable():
    if _has_systemctl():
        subprocess.run(["systemctl", "--user", "disable", "--now", LABEL + ".service"], capture_output=True)
    try:
        os.remove(_linux_unit())
    except FileNotFoundError:
        print("  (it wasn't set to start on login.)")
        return 0
    except OSError as e:
        print(f"  couldn't remove the unit: {e}")
        return 1
    if _has_systemctl():
        subprocess.run(["systemctl", "--user", "daemon-reload"], capture_output=True)
    print("  removed the systemd service — auto-join won't start on login anymore.")
    return 0


def _linux_is_enabled():
    return os.path.exists(_linux_unit())


# ── platform dispatch ───────────────────────────────────────────────────────
def enable():
    if sys.platform == "win32":
        rc = _win_enable()
        _start_now()                             # Startup only fires at login; start it now too
        return rc
    if sys.platform == "darwin":
        return _mac_enable()                     # launchd RunAtLoad starts it now
    return _linux_enable()                        # systemd --now starts it now


def disable():
    if sys.platform == "win32":
        _stop_now()
        return _win_disable()
    if sys.platform == "darwin":
        return _mac_disable()
    return _linux_disable()


def is_enabled():
    if sys.platform == "win32":
        return _win_is_enabled()
    if sys.platform == "darwin":
        return _mac_is_enabled()
    return _linux_is_enabled()


def _start_now():
    subprocess.run([sys.executable, AUTOJOIN, "start"])


def _stop_now():
    subprocess.run([sys.executable, AUTOJOIN, "stop"])


if __name__ == "__main__":
    cmd = sys.argv[1] if len(sys.argv) > 1 else ""
    if cmd == "enable":
        raise SystemExit(enable())
    if cmd == "disable":
        raise SystemExit(disable())
    if cmd == "status":
        print("start-on-login:", "yes" if is_enabled() else "no")
        raise SystemExit(0)
    print("usage: python autostart.py [enable|disable|status]")

File: meeting-notetaker/node/autostart.js
"use strict";
/*
 * Start auto-join automatically when you log in — OS-specific bit (Node mirror of
 * python/autostart.py). Native, no-admin mechanism per platform, and we print the
 * exact file created so nothing happens behind your back:
 *
 *   Windows   a hidden launcher in your Startup folder
 *   macOS     a LaunchAgent (~/Library/LaunchAgents)
 *   Linux     a systemd --user service (~/.config/systemd/user)
 *
 * enable() also starts it now; disable() also stops it now.
 */
const fs = require("fs");
const os = require("os");
const path = require("path");
const { spawnSync, execFileSync } = require("child_process");

const HERE = __dirname;
const AUTOJOIN = path.join(HERE, "autojoin.js");
const NODE = process.execPath;
const APP_NAME = "MeetingNotetaker-AutoJoin";
const LABEL = "dev.agentcall.notetaker-autojoin";
const BOOT_LOG = path.join(path.dirname(HERE), ".notetaker", "autojoin.boot.log");

// ── Windows: a hidden .vbs in the Startup folder ────────────────────────────
function winStartupDir() {
  return path.join(process.env.APPDATA || "", "Microsoft", "Windows", "Start Menu", "Programs", "Startup");
}
function winVbs() { return path.join(winStartupDir(), APP_NAME + ".vbs"); }
function winEnable() {
  const vbs = winVbs();
  const runArg = `"""${NODE}"" ""${AUTOJOIN}"" start"`;   // VBS doubles internal quotes
  const content = 'Set sh = CreateObject("WScript.Shell")\r\n' +
    `sh.CurrentDirectory = "${HERE}"\r\n` +
    `sh.Run ${runArg}, 0, False\r\n`;
  fs.mkdirSync(winStartupDir(), { recursive: true });
  fs.writeFileSync(vbs, content);
  console.log("  ✓ added a startup launcher so auto-join starts when you log in:");
  console.log(`    ${vbs}`);
  return 0;
}
function winDisable() {
  try { fs.unlinkSync(winVbs()); console.log("  removed the startup launcher — auto-join won't start on login anymore."); }
  catch (e) { if (e.code === "ENOENT") console.log("  (it wasn't set to start on login.)"); else { console.log(`  couldn't remove ${winVbs()}: ${e.message}`); return 1; } }
  return 0;
}
function winIsEnabled() { try { return fs.existsSync(winVbs()); } catch { return false; } }

// ── macOS: a LaunchAgent plist ──────────────────────────────────────────────
function macPlist() { return path.join(os.homedir(), "Library", "LaunchAgents", LABEL + ".plist"); }
function macEnable() {
  const plist = macPlist();
  fs.mkdirSync(path.dirname(plist), { recursive: true });
  const xml = `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key><string>${LABEL}</string>
  <key>ProgramArguments</key>
  <array><string>${NODE}</string><string>${AUTOJOIN}</string><string>run</string></array>
  <key>WorkingDirectory</key><string>${HERE}</string>
  <key>RunAtLoad</key><true/>
  <key>KeepAlive</key><dict><key>SuccessfulExit</key><false/></dict>
  <key>StandardOutPath</key><string>${BOOT_LOG}</string>
  <key>StandardErrorPath</key><string>${BOOT_LOG}</string>
</dict>
</plist>
`;
  fs.writeFileSync(plist, xml);
  spawnSync("launchctl", ["unload", plist]);
  const r = spawnSync("launchctl", ["load", "-w", plist], { encoding: "utf-8" });
  console.log(`  ✓ installed a LaunchAgent: ${plist}`);
  if (r.status !== 0) {
    console.log(`  (launchctl load said: ${(r.stderr || r.stdout || "").trim()})`);
    console.log(`  You can load it yourself with:  launchctl load -w "${plist}"`);
  }
  return 0;
}
function macDisable() {
  const plist = macPlist();
  spawnSync("launchctl", ["unload", "-w", plist]);
  try { fs.unlinkSync(plist); } catch (e) { if (e.code === "ENOENT") { console.log("  (it wasn't set to start on login.)"); return 0; } console.log(`  couldn't remove ${plist}: ${e.message}`); return 1; }
  console.log("  removed the LaunchAgent — auto-join won't start on login anymore.");
  return 0;
}
function macIsEnabled() { try { return fs.existsSync(macPlist()); } catch { return false; } }

// ── Linux: a systemd --user service ─────────────────────────────────────────
function linuxUnit() { return path.join(os.homedir(), ".config", "systemd", "user", LABEL + ".service"); }
function hasSystemctl() { try { execFileSync("systemctl", ["--version"], { stdio: "ignore" }); return true; } catch { return false; } }
function linuxEnable() {
  if (!hasSystemctl()) {
    console.log("  systemd isn't available here. Add this to your login startup by hand:");
    console.log(`    ${NODE} ${AUTOJOIN} start`);
    return 1;
  }
  const unit = linuxUnit();
  fs.mkdirSync(path.dirname(unit), { recursive: true });
  fs.writeFileSync(unit,
    "[Unit]\nDescription=Meeting Notetaker auto-join (watches your calendar)\n\n" +
    `[Service]\nWorkingDirectory=${HERE}\nExecStart=${NODE} ${AUTOJOIN} run\nRestart=on-failure\n\n` +
    "[Install]\nWantedBy=default.target\n");
  spawnSync("systemctl", ["--user", "daemon-reload"]);
  const r = spawnSync("systemctl", ["--user", "enable", "--now", LABEL + ".service"], { encoding: "utf-8" });
  console.log(`  ✓ installed a systemd --user service: ${unit}`);
  if (r.status !== 0) console.log(`  (systemctl said: ${(r.stderr || r.stdout || "").trim()})`);
  console.log("  Tip: to run it even before you log in, enable lingering once:");
  console.log("    loginctl enable-linger $USER");
  return 0;
}
function linuxDisable() {
  if (hasSystemctl()) spawnSync("systemctl", ["--user", "disable", "--now", LABEL + ".service"]);
  try { fs.unlinkSync(linuxUnit()); } catch (e) { if (e.code === "ENOENT") { console.log("  (it wasn't set to start on login.)"); return 0; } console.log(`  couldn't remove the unit: ${e.message}`); return 1; }
  if (hasSystemctl()) spawnSync("systemctl", ["--user", "daemon-reload"]);
  console.log("  removed the systemd service — auto-join won't start on login anymore.");
  return 0;
}
function linuxIsEnabled() { try { return fs.existsSync(linuxUnit()); } catch { return false; } }

// ── dispatch ─────────────────────────────────────────────────────────────────
function startNow() { spawnSync(NODE, [AUTOJOIN, "start"], { stdio: "inherit" }); }
function stopNow() { spawnSync(NODE, [AUTOJOIN, "stop"], { stdio: "inherit" }); }

function enable() {
  if (process.platform === "win32") { const rc = winEnable(); startNow(); return rc; }
  if (process.platform === "darwin") return macEnable();
  return linuxEnable();
}
function disable() {
  if (process.platform === "win32") { stopNow(); return winDisable(); }
  if (process.platform === "darwin") return macDisable();
  return linuxDisable();
}
function isEnabled() {
  if (process.platform === "win32") return winIsEnabled();
  if (process.platform === "darwin") return macIsEnabled();
  return linuxIsEnabled();
}

module.exports = { enable, disable, isEnabled };

if (require.main === module) {
  const cmd = process.argv[2];
  if (cmd === "enable") process.exit(enable());
  else if (cmd === "disable") process.exit(disable());
  else if (cmd === "status") { console.log("start-on-login:", isEnabled() ? "yes" : "no"); process.exit(0); }
  else { console.log("usage: node autostart.js [enable|disable|status]"); process.exit(0); }
}
